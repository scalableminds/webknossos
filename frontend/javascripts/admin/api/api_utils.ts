import Toast from "libs/toast";
import messages from "messages";

const MAX_SERVER_ITEMS_PER_RESPONSE = 1000;

export function assertResponseLimit(collection: unknown[]) {
  if (collection.length === MAX_SERVER_ITEMS_PER_RESPONSE) {
    Toast.warning(messages["request.max_item_count_alert"], {
      sticky: true,
    });
  }
}
