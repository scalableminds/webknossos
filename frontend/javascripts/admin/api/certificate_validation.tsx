import Request from "libs/request";

type CertificateValidationResult = {
  isValid: boolean;
  expiresAt: number;
};

export async function isCertificateValid(): Promise<CertificateValidationResult> {
  try {
    return await Request.receiveJSON("/api/checkCertificate", { showErrorToast: false });
  } catch (errorResponse: any) {
    if (errorResponse.status !== 400) {
      // In case the server is not available or some other kind of error occurred, we assume the certificate is valid.
      return { isValid: true, expiresAt: 0 };
    }
    try {
      const { isValid, expiresAt } = JSON.parse(errorResponse.errors[0]);
      return { isValid, expiresAt };
    } catch (_e) {
      // If parsing the error message fails, we assume the certificate is valid.
      return { isValid: true, expiresAt: 0 };
    }
  }
}
