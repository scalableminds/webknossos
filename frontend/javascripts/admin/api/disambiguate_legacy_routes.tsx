import Request, { type RequestOptions } from "libs/request";

export async function getOrganizationForDataset(datasetName: string): Promise<string> {
  const { organization: organizationId } = await Request.receiveJSON(
    `/api/datasets/disambiguate/${datasetName}/toNew`,
  );
  return organizationId;
}

export async function getDatasetIdFromNameAndOrganization(
  datasetName: string,
  organizationId: string,
  sharingToken?: string | null | undefined,
  options: RequestOptions = {},
): Promise<string> {
  const sharingTokenSuffix = sharingToken != null ? `?sharingToken=${sharingToken}` : "";
  const { id: datasetId } = await Request.receiveJSON(
    `/api/datasets/disambiguate/${organizationId}/${datasetName}/toId${sharingTokenSuffix}`,
    options,
  );
  return datasetId;
}
