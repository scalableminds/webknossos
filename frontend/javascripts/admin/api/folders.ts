import Request from "libs/request";
import { Folder, FlatFolderTreeItem, FolderUpdater } from "types/api_flow_types";

export function getFolder(folderId: string): Promise<Folder> {
  return Request.receiveJSON(`/api/folders/${folderId}`);
}

export function getFolderTree(): Promise<FlatFolderTreeItem[]> {
  return Request.receiveJSON("/api/folders/tree");
}

export async function createFolder(parentId: string, name: string): Promise<Folder> {
  const params = new URLSearchParams();
  params.append("parentId", parentId);
  params.append("name", name);

  const folder = await Request.receiveJSON(`/api/folders/create?${params}`, {
    method: "POST",
  });
  return folder;
}

export async function deleteFolder(id: string): Promise<string> {
  await Request.receiveJSON(`/api/folders/${id}`, {
    method: "DELETE",
  });
  // Return deleted id
  return id;
}

export function updateFolder(folder: FolderUpdater): Promise<Folder> {
  return Request.sendJSONReceiveJSON(`/api/folders/${folder.id}`, {
    data: folder,
    method: "PUT",
  });
}

export function moveFolder(folderId: string, newParentId: string): Promise<Folder> {
  const params = new URLSearchParams();
  params.append("newParentId", newParentId);
  return Request.sendJSONReceiveJSON(`/api/folders/${folderId}/move?${params}`, {
    data: {},
    method: "PUT",
  });
}
