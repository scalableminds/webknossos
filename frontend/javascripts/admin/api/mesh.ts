import Request from "libs/request";
import type { APIDataSourceId, APIMeshFileInfo } from "types/api_types";
import type { Vector3, Vector4 } from "viewer/constants";
import { doWithToken } from "./token";

export type MeshChunk = {
  position: Vector3;
  byteOffset: number;
  byteSize: number;
  unmappedSegmentId: number;
};

export type MeshLodInfo = {
  chunks: Array<MeshChunk>;
  transform: [Vector4, Vector4, Vector4]; // 4x3 matrix
};

type MeshSegmentInfo = {
  meshFormat: "draco";
  lods: Array<MeshLodInfo>;
  chunkScale: Vector3;
};

type ListMeshChunksRequest = {
  meshFile: APIMeshFileInfo;
  segmentId: number;
};

export function getMeshfileChunksForSegment(
  dataStoreUrl: string,
  dataSourceId: APIDataSourceId,
  layerName: string,
  meshFile: APIMeshFileInfo,
  segmentId: number,
  // targetMappingName is the on-disk mapping name.
  // In case of an editable mapping, this should still be the on-disk base
  // mapping name (so that agglomerates that are untouched by the editable
  // mapping can be looked up there without another round-trip between tracingstore
  // and datastore)
  targetMappingName: string | null | undefined,
  // editableMappingTracingId should be the tracing id, not the editable mapping id.
  // If this is set, it is assumed that the request is about an editable mapping.
  editableMappingTracingId: string | null | undefined,
): Promise<MeshSegmentInfo> {
  return doWithToken((token) => {
    const params = new URLSearchParams();
    params.append("token", token);
    if (targetMappingName != null) {
      params.append("targetMappingName", targetMappingName);
    }
    if (editableMappingTracingId != null) {
      params.append("editableMappingTracingId", editableMappingTracingId);
    }
    const payload: ListMeshChunksRequest = {
      meshFile.name,
      segmentId,
    };
    return Request.sendJSONReceiveJSON(
      `${dataStoreUrl}/data/datasets/${dataSourceId.owningOrganization}/${dataSourceId.directoryName}/layers/${layerName}/meshes/chunks?${params}`,
      {
        data: payload,
        showErrorToast: false,
      },
    );
  });
}

type MeshChunkDataRequest = {
  byteOffset: number;
  byteSize: number;
  segmentId: number | null; // Only relevant for neuroglancer precomputed meshes
};

type MeshChunkDataRequestList = {
  meshFileName: string;
  requests: MeshChunkDataRequest[];
};

export function getMeshfileChunkData(
  dataStoreUrl: string,
  dataSourceId: APIDataSourceId,
  layerName: string,
  batchDescription: MeshChunkDataRequestList,
): Promise<ArrayBuffer[]> {
  return doWithToken(async (token) => {
    const dracoDataChunks = await Request.sendJSONReceiveArraybuffer(
      `${dataStoreUrl}/data/datasets/${dataSourceId.owningOrganization}/${dataSourceId.directoryName}/layers/${layerName}/meshes/chunks/data?token=${token}`,
      {
        data: batchDescription,
        useWebworkerForArrayBuffer: true,
      },
    );
    const chunkCount = batchDescription.requests.length;
    const jumpPositionsForChunks = [];
    let cumsum = 0;
    for (const req of batchDescription.requests) {
      jumpPositionsForChunks.push(cumsum);
      cumsum += req.byteSize;
    }
    jumpPositionsForChunks.push(cumsum);

    const dataEntries = [];
    for (let chunkIdx = 0; chunkIdx < chunkCount; chunkIdx++) {
      // slice() creates a copy of the data, but working with TypedArray Views would cause
      // issues when transferring the data to a webworker.
      const dracoData = dracoDataChunks.slice(
        jumpPositionsForChunks[chunkIdx],
        jumpPositionsForChunks[chunkIdx + 1],
      );
      dataEntries.push(dracoData);
    }

    return dataEntries;
  });
}
