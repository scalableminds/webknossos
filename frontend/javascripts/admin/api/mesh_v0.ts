import Request from "libs/request";
import { Vector3 } from "oxalis/constants";
import { APIDatasetId } from "types/api_flow_types";
import { doWithToken } from "./token";

export function getMeshfileChunksForSegment(
  dataStoreUrl: string,
  datasetId: APIDatasetId,
  layerName: string,
  meshFile: string,
  segmentId: number,
): Promise<Array<Vector3>> {
  return doWithToken((token) => {
    const params = new URLSearchParams();
    params.append("token", token);
    return Request.sendJSONReceiveJSON(
      `${dataStoreUrl}/data/datasets/${datasetId.owningOrganization}/${datasetId.name}/layers/${layerName}/meshes/chunks?${params}`,
      {
        data: {
          meshFile,
          segmentId,
        },
        showErrorToast: false,
      },
    );
  });
}

export function getMeshfileChunkData(
  dataStoreUrl: string,
  datasetId: APIDatasetId,
  layerName: string,
  meshFile: string,
  segmentId: number,
  position: Vector3,
): Promise<ArrayBuffer> {
  return doWithToken(async (token) => {
    const data = await Request.sendJSONReceiveArraybufferWithHeaders(
      `${dataStoreUrl}/data/datasets/${datasetId.owningOrganization}/${datasetId.name}/layers/${layerName}/meshes/chunks/data?token=${token}`,
      {
        data: {
          meshFile,
          segmentId,
          position,
        },
        useWebworkerForArrayBuffer: false,
      },
    );
    return data;
  });
}
