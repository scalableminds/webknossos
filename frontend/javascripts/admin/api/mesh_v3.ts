import Request from "libs/request";
import _ from "lodash";
import { Vector3, Vector4 } from "oxalis/constants";
import { APIDatasetId } from "types/api_flow_types";
import { doWithToken } from "./token";

export type MeshChunk = { position: Vector3; byteOffset: number; byteSize: number };

type MeshLodInfo = {
  scale: number;
  vertexOffset: Vector3;
  chunkShape: Vector3;
  chunks: Array<MeshChunk>;
};

type MeshSegmentInfo = {
  chunkShape: Vector3;
  gridOrigin: Vector3;
  lods: Array<MeshLodInfo>;
};

type SegmentInfo = {
  transform: [Vector4, Vector4, Vector4]; // 4x3 matrix
  meshFormat: "draco";
  chunks: MeshSegmentInfo;
};

export function getMeshfileChunksForSegment(
  dataStoreUrl: string,
  datasetId: APIDatasetId,
  layerName: string,
  meshFile: string,
  segmentId: number,
  // targetMappingName is the on-disk mapping name.
  // In case of an editable mapping, this should still be the on-disk base
  // mapping name (so that agglomerates that are untouched by the editable
  // mapping can be looked up there without another round-trip between tracingstore
  // and datastore)
  targetMappingName: string | null | undefined,
  // editableMappingTracingId should be the tracing id, not the editable mapping id.
  // If this is set, it is assumed that the request is about an editable mapping.
  editableMappingTracingId: string | null | undefined,
): Promise<SegmentInfo> {
  return doWithToken((token) => {
    const params = new URLSearchParams();
    params.append("token", token);
    if (targetMappingName != null) {
      params.append("targetMappingName", targetMappingName);
    }
    if (editableMappingTracingId != null) {
      params.append("editableMappingTracingId", editableMappingTracingId);
    }
    return Request.sendJSONReceiveJSON(
      `${dataStoreUrl}/data/datasets/${datasetId.owningOrganization}/${datasetId.name}/layers/${layerName}/meshes/formatVersion/3/chunks?${params}`,
      {
        data: {
          meshFile,
          segmentId,
        },
        showErrorToast: false,
      },
    );
  });
}

type MeshChunkDataRequestV3 = {
  byteOffset: number;
  byteSize: number;
};

type MeshChunkDataRequestV3List = {
  meshFile: string;
  requests: MeshChunkDataRequestV3[];
};

export function getMeshfileChunkData(
  dataStoreUrl: string,
  datasetId: APIDatasetId,
  layerName: string,
  batchDescription: MeshChunkDataRequestV3List,
): Promise<ArrayBuffer[]> {
  return doWithToken(async (token) => {
    const dracoDataChunks = await Request.sendJSONReceiveArraybuffer(
      `${dataStoreUrl}/data/datasets/${datasetId.owningOrganization}/${datasetId.name}/layers/${layerName}/meshes/formatVersion/3/chunks/data?token=${token}`,
      {
        data: batchDescription,
        useWebworkerForArrayBuffer: true,
      },
    );
    const chunkCount = batchDescription.requests.length;
    const jumpPositionsForChunks = [];
    let cumsum = 0;
    for (const req of batchDescription.requests) {
      jumpPositionsForChunks.push(cumsum);
      cumsum += req.byteSize;
    }
    jumpPositionsForChunks.push(cumsum);

    const dataEntries = [];
    for (let chunkIdx = 0; chunkIdx < chunkCount; chunkIdx++) {
      // slice() creates a copy of the data, but working with TypedArray Views would cause
      // issues when transferring the data to a webworker.
      const dracoData = dracoDataChunks.slice(
        jumpPositionsForChunks[chunkIdx],
        jumpPositionsForChunks[chunkIdx + 1],
      );
      dataEntries.push(dracoData);
    }

    return dataEntries;
  });
}
