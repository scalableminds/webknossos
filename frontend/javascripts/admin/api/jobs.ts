import Request from "libs/request";
import { location } from "libs/window";
import type { UnitLong, Vector3, Vector6 } from "oxalis/constants";
import type {
  APIAnnotationType,
  APIJob,
  APIJobState,
  APIJobManualState,
  APIEffectiveJobState,
  AiModel,
  RenderAnimationOptions,
} from "types/api_flow_types";
import { assertResponseLimit } from "./api_utils";

function transformBackendJobToAPIJob(job: any): APIJob {
  return {
    id: job.id,
    type: job.command,
    datasetName: job.commandArgs.dataset_name,
    organizationName: job.commandArgs.organization_name,
    layerName: job.commandArgs.layer_name || job.commandArgs.volume_layer_name,
    annotationLayerName: job.commandArgs.annotation_layer_name,
    boundingBox: job.commandArgs.bbox,
    exportFileName: job.commandArgs.export_file_name,
    tracingId: job.commandArgs.volume_tracing_id,
    annotationId: job.commandArgs.annotation_id,
    annotationType: job.commandArgs.annotation_type,
    mergeSegments: job.commandArgs.merge_segments,
    trainingAnnotations: job.commandArgs.training_annotations,
    state: adaptJobState(job.state, job.manualState),
    manualState: job.manualState,
    result: job.returnValue,
    resultLink: job.resultLink,
    createdAt: job.created,
    voxelyticsWorkflowHash: job.voxelyticsWorkflowHash,
  };
}

export async function getJobs(): Promise<APIJob[]> {
  const jobs = await Request.receiveJSON("/api/jobs");
  assertResponseLimit(jobs);
  return (
    jobs
      .map(transformBackendJobToAPIJob)
      // Newest jobs should be first
      .sort((a: APIJob, b: APIJob) => a.createdAt > b.createdAt)
  );
}

export async function getJob(jobId: string): Promise<APIJob> {
  const job = await Request.receiveJSON(`/api/jobs/${jobId}`);
  return transformBackendJobToAPIJob(job);
}

function adaptJobState(
  celeryState: APIJobState,
  manualState: APIJobManualState,
): APIEffectiveJobState {
  if (manualState) {
    return manualState;
  }

  return celeryState || "UNKNOWN";
}

export async function cancelJob(jobId: string): Promise<APIJob> {
  return Request.receiveJSON(`/api/jobs/${jobId}/cancel`, {
    method: "PATCH",
  });
}

export async function startConvertToWkwJob(
  datasetName: string,
  organizationName: string,
  scale: Vector3,
  unit: UnitLong,
): Promise<APIJob> {
  return Request.receiveJSON(
    `/api/jobs/run/convertToWkw/${organizationName}/${datasetName}?scale=${scale.toString()}&unit=${unit}`,
    {
      method: "POST",
    },
  );
}

export async function startFindLargestSegmentIdJob(
  datasetName: string,
  organizationName: string,
  layerName: string,
): Promise<APIJob> {
  return Request.receiveJSON(
    `/api/jobs/run/findLargestSegmentId/${organizationName}/${datasetName}?layerName=${layerName}`,
    {
      method: "POST",
    },
  );
}

export async function startExportTiffJob(
  datasetName: string,
  organizationName: string,
  bbox: Vector6,
  layerName: string | null | undefined,
  mag: string | null | undefined,
  annotationId: string | null | undefined,
  annotationLayerName: string | null | undefined,
  asOmeTiff: boolean,
): Promise<APIJob> {
  const params = new URLSearchParams({ bbox: bbox.join(","), asOmeTiff: asOmeTiff.toString() });
  if (layerName != null) {
    params.append("layerName", layerName);
  }
  if (mag != null) {
    params.append("mag", mag);
  }
  if (annotationId != null) {
    params.append("annotationId", annotationId);
  }
  if (annotationLayerName != null) {
    params.append("annotationLayerName", annotationLayerName);
  }
  return Request.receiveJSON(
    `/api/jobs/run/exportTiff/${organizationName}/${datasetName}?${params}`,
    {
      method: "POST",
    },
  );
}

export function startComputeMeshFileJob(
  organizationName: string,
  datasetName: string,
  layerName: string,
  mag: Vector3,
  agglomerateView?: string,
): Promise<APIJob> {
  const params = new URLSearchParams();
  params.append("layerName", layerName);
  params.append("mag", mag.join("-"));

  if (agglomerateView) {
    params.append("agglomerateView", agglomerateView);
  }

  return Request.receiveJSON(
    `/api/jobs/run/computeMeshFile/${organizationName}/${datasetName}?${params}`,
    {
      method: "POST",
    },
  );
}

export function startComputeSegmentIndexFileJob(
  organizationName: string,
  datasetName: string,
  layerName: string,
): Promise<APIJob> {
  const params = new URLSearchParams();
  params.append("layerName", layerName);

  return Request.receiveJSON(
    `/api/jobs/run/computeSegmentIndexFile/${organizationName}/${datasetName}?${params}`,
    {
      method: "POST",
    },
  );
}

export function startNucleiInferralJob(
  organizationName: string,
  datasetName: string,
  layerName: string,
  newDatasetName: string,
): Promise<APIJob> {
  return Request.receiveJSON(
    `/api/jobs/run/inferNuclei/${organizationName}/${datasetName}?layerName=${layerName}&newDatasetName=${newDatasetName}`,
    {
      method: "POST",
    },
  );
}

export function startNeuronInferralJob(
  organizationName: string,
  datasetName: string,
  layerName: string,
  bbox: Vector6,
  outputSegmentationLayerName: string,
  newDatasetName: string,
): Promise<APIJob> {
  const urlParams = new URLSearchParams({
    layerName,
    bbox: bbox.join(","),
    outputSegmentationLayerName,
    newDatasetName,
  });
  return Request.receiveJSON(
    `/api/jobs/run/inferNeurons/${organizationName}/${datasetName}?${urlParams.toString()}`,
    {
      method: "POST",
    },
  );
}

export function startRenderAnimationJob(
  organizationName: string,
  datasetName: string,
  animationOptions: RenderAnimationOptions,
): Promise<APIJob> {
  return Request.sendJSONReceiveJSON(
    `/api/jobs/run/renderAnimation/${organizationName}/${datasetName}`,
    {
      data: animationOptions,
    },
  );
}

function startSegmentationAnnotationDependentJob(
  jobURLPath: string,
  organizationName: string,
  datasetName: string,
  fallbackLayerName: string,
  volumeLayerName: string | null | undefined,
  newDatasetName: string,
  annotationId: string,
  annotationType: APIAnnotationType,
  outputSegmentationLayerName?: string,
  mergeSegments?: boolean,
): Promise<APIJob> {
  const requestURL = new URL(
    `/api/jobs/run/${jobURLPath}/${organizationName}/${datasetName}`,
    location.origin,
  );
  if (volumeLayerName != null) {
    requestURL.searchParams.append("volumeLayerName", volumeLayerName);
  }
  requestURL.searchParams.append("fallbackLayerName", fallbackLayerName);
  requestURL.searchParams.append("annotationId", annotationId);
  requestURL.searchParams.append("annotationType", annotationType);
  requestURL.searchParams.append("newDatasetName", newDatasetName);
  if (outputSegmentationLayerName != null) {
    requestURL.searchParams.append("outputSegmentationLayerName", outputSegmentationLayerName);
  }
  if (mergeSegments != null) {
    requestURL.searchParams.append("mergeSegments", mergeSegments.toString());
  }
  return Request.receiveJSON(requestURL.href, {
    method: "POST",
  });
}

export function startMaterializingVolumeAnnotationJob(
  organizationName: string,
  datasetName: string,
  fallbackLayerName: string,
  volumeLayerName: string | null | undefined,
  newDatasetName: string,
  outputSegmentationLayerName: string,
  annotationId: string,
  annotationType: APIAnnotationType,
  mergeSegments: boolean,
): Promise<APIJob> {
  return startSegmentationAnnotationDependentJob(
    "materializeVolumeAnnotation",
    organizationName,
    datasetName,
    fallbackLayerName,
    volumeLayerName,
    newDatasetName,
    annotationId,
    annotationType,
    outputSegmentationLayerName,
    mergeSegments,
  );
}

export function startMitochondriaInferralJob(
  organizationName: string,
  datasetName: string,
  layerName: string,
  bbox: Vector6,
  outputSegmentationLayerName: string,
  newDatasetName: string,
): Promise<APIJob> {
  const urlParams = new URLSearchParams({
    layerName,
    bbox: bbox.join(","),
    outputSegmentationLayerName,
    newDatasetName,
  });
  return Request.receiveJSON(
    `/api/jobs/run/inferMitochondria/${organizationName}/${datasetName}?${urlParams.toString()}`,
    {
      method: "POST",
    },
  );
}

export function startAlignSectionsJob(
  organizationName: string,
  datasetName: string,
  layerName: string,
  newDatasetName: string,
): Promise<APIJob> {
  const urlParams = new URLSearchParams({
    layerName,
    newDatasetName,
  });
  return Request.receiveJSON(
    `/api/jobs/run/alignSections/${organizationName}/${datasetName}?${urlParams.toString()}`,
    {
      method: "POST",
    },
  );
}

type AiModelCategory = "em_neurons" | "em_nuclei";

type AiModelTrainingAnnotationSpecification = {
  annotationId: string;
  colorLayerName: string;
  segmentationLayerName: string;
  mag: Vector3;
};

type RunTrainingParameters = {
  trainingAnnotations: Array<AiModelTrainingAnnotationSpecification>;
  name: string;
  comment?: string;
  aiModelCategory?: AiModelCategory;
  workflowYaml?: string;
};

export function runTraining(params: RunTrainingParameters) {
  return Request.sendJSONReceiveJSON("/api/aiModels/runTraining", {
    method: "POST",
    data: JSON.stringify(params),
  });
}

type RunInferenceParameters = {
  annotationId?: string;
  aiModelId: string;
  datasetName: string;
  colorLayerName: string;
  boundingBox: Vector6;
  newSegmentationLayerName: string;
  newDatasetName: string;
  // maskAnnotationLayerName?: string | null
};

export function runInferenceJob(params: RunInferenceParameters) {
  return Request.sendJSONReceiveJSON("/api/aiModels/inferences/runInference", {
    method: "POST",
    data: JSON.stringify({ ...params, boundingBox: params.boundingBox.join(",") }),
  });
}

export async function getAiModels(): Promise<AiModel[]> {
  const models = await Request.receiveJSON("/api/aiModels");
  return models.map((model: any) => ({
    ...model,
    trainingJob: model.trainingJob == null ? null : transformBackendJobToAPIJob(model.trainingJob),
  }));
}
