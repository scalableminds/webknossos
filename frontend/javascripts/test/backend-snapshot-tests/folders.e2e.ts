import _ from "lodash";
import {
  tokenUserA,
  tokenUserC,
  setCurrToken,
  resetDatabase,
  replaceVolatileValues,
  writeTypeCheckingFile,
} from "test/enzyme/e2e-setup";
import Request from "libs/request";
import * as foldersApi from "admin/api/folders";
import test from "ava";
test.before("Reset database and change token", async () => {
  resetDatabase();
  setCurrToken(tokenUserA);
});
test("getFolderTree", async (t) => {
  const folderTree = _.sortBy(
    await foldersApi.getFolderTree(),
    (folderWithParent) => folderWithParent.name,
  );

  writeTypeCheckingFile(folderTree, "folderTree", "FlatFolderTreeItem", {
    isArray: true,
  });
  t.snapshot(folderTree, {
    id: "folders-getFolderTree()",
  });
});
const organizationXRootFolderId = "570b9f4e4bb848d0885ea917";
test("getFolder", async (t) => {
  const folder = await foldersApi.getFolder(organizationXRootFolderId);

  writeTypeCheckingFile(folder, "folder", "Folder");
  t.snapshot(folder, {
    id: "folders-getFolder()",
  });
});
test("updateFolder", async (t) => {
  const newName = "renamed organization x root folder";
  const updatedFolder = await foldersApi.updateFolder({
    id: organizationXRootFolderId,
    allowedTeams: [],
    name: newName,
  });
  t.is(updatedFolder.name, newName);

  t.snapshot(updatedFolder, {
    id: "folders-updatedFolder()",
  });
});
test("createFolder", async (t) => {
  const newName = "a newly created folder!";
  const folder = await foldersApi.createFolder(organizationXRootFolderId, newName);
  t.is(folder.name, newName);

  t.snapshot(replaceVolatileValues(folder), {
    id: "folders-createFolder()",
  });
});
test("addAllowedTeamToFolder", async (t) => {
  const subFolderId = "570b9f4e4bb848d08880712a";
  const anotherSubFolderId = "570b9f4e4bb848d08880712b";
  const teamId = "570b9f4b2a7c0e3b008da6ec";

  await Request.receiveJSON(`/api/folders/${subFolderId}`);

  const updatedFolderWithTeam = await foldersApi.updateFolder({
    id: subFolderId,
    allowedTeams: [teamId],
    name: "A subfolder!",
  });

  t.snapshot(updatedFolderWithTeam, {
    id: "folders-updatedWithTeam",
  });

  setCurrToken(tokenUserC);
  /*
   * user c does not have elevated permissions, but is a member of the team
   * we expect that they can see the subfolder we add the team to, but not the other.
   */

  const subFolderSeenByUserC = await foldersApi.getFolder(subFolderId);

  t.snapshot(subFolderSeenByUserC, {
    id: "folders-folderAfterUpdateTeamsSeenByUserC",
  });

  await t.throwsAsync(async () => {
    try {
      await Request.receiveJSON(`/api/folders/${anotherSubFolderId}`);
    } catch {
      throw new Error("request failed");
    }
  });
});
