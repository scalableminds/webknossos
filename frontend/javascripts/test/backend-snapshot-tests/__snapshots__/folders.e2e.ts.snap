// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Folder API (E2E) > addAllowedTeamToFolder 1`] = `
{
  "allowedTeams": [
    {
      "id": "570b9f4b2a7c0e3b008da6ec",
      "name": "team_X1",
      "organization": "Organization_X",
    },
  ],
  "allowedTeamsCumulative": [
    {
      "id": "570b9f4b2a7c0e3b008da6ec",
      "name": "team_X1",
      "organization": "Organization_X",
    },
  ],
  "id": "570b9f4e4bb848d08880712a",
  "isEditable": true,
  "metadata": [
    {
      "key": "foo",
      "type": "string",
      "value": "bar",
    },
  ],
  "name": "A subfolder!",
}
`;

exports[`Folder API (E2E) > addAllowedTeamToFolder 2`] = `
{
  "allowedTeams": [
    {
      "id": "570b9f4b2a7c0e3b008da6ec",
      "name": "team_X1",
      "organization": "Organization_X",
    },
  ],
  "allowedTeamsCumulative": [
    {
      "id": "570b9f4b2a7c0e3b008da6ec",
      "name": "team_X1",
      "organization": "Organization_X",
    },
  ],
  "id": "570b9f4e4bb848d08880712a",
  "isEditable": false,
  "metadata": [
    {
      "key": "foo",
      "type": "string",
      "value": "bar",
    },
  ],
  "name": "A subfolder!",
}
`;

exports[`Folder API (E2E) > createFolder 1`] = `
{
  "allowedTeams": [],
  "allowedTeamsCumulative": [],
  "id": "id",
  "isEditable": true,
  "metadata": [],
  "name": "a newly created folder!",
}
`;

exports[`Folder API (E2E) > getFolder 1`] = `
{
  "allowedTeams": [],
  "allowedTeamsCumulative": [],
  "id": "570b9f4e4bb848d0885ea917",
  "isEditable": true,
  "metadata": [
    {
      "key": "key",
      "type": "number",
      "value": 10,
    },
  ],
  "name": "Organization_X",
}
`;

exports[`Folder API (E2E) > getFolderTree 1`] = `
[
  {
    "id": "570b9f4e4bb848d08880712a",
    "isEditable": true,
    "metadata": [],
    "name": "A subfolder!",
    "parent": "570b9f4e4bb848d0885ea917",
  },
  {
    "id": "570b9f4e4bb848d08880712b",
    "isEditable": true,
    "metadata": [],
    "name": "Another subfolder!",
    "parent": "570b9f4e4bb848d0885ea917",
  },
  {
    "id": "570b9f4e4bb848d0885ea917",
    "isEditable": true,
    "metadata": [
      {
        "key": "key",
        "type": "number",
        "value": 10,
      },
    ],
    "name": "Organization_X",
    "parent": null,
  },
]
`;

exports[`Folder API (E2E) > updateFolder 1`] = `
{
  "allowedTeams": [],
  "allowedTeamsCumulative": [],
  "id": "570b9f4e4bb848d0885ea917",
  "isEditable": true,
  "metadata": [],
  "name": "renamed organization x root folder",
}
`;
